Add tracker's error reporting (close #343)

AlexBenny opened the pull request:
This feature is useful to track errors raised by the tracker to the collector as `diagnostic_error` events.
Previously the tracker treated internal errors managing them silently. They were visible in the logs of the app available on debugging only.
This can be particularly useful when an issue is not reproducible in debug mode but is visible on production. The user can collect and filter diagnostic_error events and report them to the support team if needed.

**Here the proposed documentation to review:**

---

### Tracker's diagnostic

The tracker is constantly tested with different platforms and versions but there are rare situations where an error can be reported. The tracker can manage the situation avoiding crashes of the app and assuring that the events are not lost. However, it can report errors to the collector as `diagnostic_error` events.

To activate this feature you only need to enable `trackerDiagnostic`:

```
TrackerBuilder trackerBuilder =
    new TrackerBuilder(emitter, namespace, appId, appContext)
        .trackerDiagnostic(true)
        ...
        .build();
Tracker.init(trackerBuilder);
```

paulboocock commented on snowplow-demo-app/src/main/java/com/snowplowanalytics/snowplowtrackerdemo/ErrorLogger.java:
```diff
@@ -0,0 +1,14 @@
+    @Override
+    public void log(String source, String errorMessage) {
+        if (source.equals("Tracker") || source.equals("Session") || source.equals("ProcessObserver")) {
+            Log.e("Error reported >>> " + source, errorMessage);
```
Maybe worth adding a Tracking example here, send the error message into the collector? Just a thought but not a blocker.

AlexBenny commented:
@paulboocock I've started working on an example in the demo app with the purpose to use the tracker to track back possible errors as `application_error` as you suggested. But it's a bit dangerous because the previous implementation may generate infinite loops. An event can generate an error that generate another event and so on.

At the moment I've removed the risk allowing error tracking only for specific errors and excluding to add contexts (especially the Session one) to error events not important for our purpose.

Now the user app doesn't need to implement any delegate. The tracker is able to send error events autonomously. The user needs only to activate the flag `trackerDiagnostic` on setup of the Tracker.

Please, could you to confirm that I've correctly implemented the payload for the TrackerError events and sending it correctly? I've done some tests and it seems ok to me but I'd like to have your double check. That should be use to spot issues in the tracker along (and in sync) with the stream of events collected.

paulboocock commented on snowplow-demo-app/src/main/java/com/snowplowanalytics/snowplowtrackerdemo/Demo.java:
```diff
@@ -315,6 +317,7 @@ private void initAndroidTracker() {
                 .mobileContext(true)
                 .geoLocationContext(true)
                 .applicationCrash(true)
+                .trackerDiagnostic(true)
```
I like this, much cleaner.

paulboocock commented on snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/tracker/ErrorLogging.java:
```diff
@@ -0,0 +1,14 @@
+package com.snowplowanalytics.snowplow.tracker;
+
+public interface ErrorLogging {
```
Should this be called DiagnosticLogger? I'm not sure it is clear if this is related ApplicationCrash or TrackerDianogistic logging.

AlexBenny replied:
Changed to DiagnosticLogger

paulboocock commented on snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/tracker/events/TrackerError.java:
```diff
@@ -0,0 +1,99 @@
+
+    @Override
+    public Payload getPayload() {
+        return new SelfDescribingJson(TrackerConstants.APPLICATION_ERROR_SCHEMA, getData());
```
I'm in two minds as to whether this is the correct schema for this or if we should create a new one.
The application_crash schema was created to report when applications that contain the tracker throw unhandled exceptions, however in this case we are allowing customers to turn on diagnostics that allow the tracker to emit log events - the application hasn't crashed at all and this could inflate users crash statistics if they are counting this event type.
I wonder if a new iglu central schema would be better and more descriptive here, perhaps `tracker_log` or `tracker_diagnostic`?

AlexBenny replied:
I would go with `tracker_diagnostic` 👍

paulboocock commented on snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/tracker/tracker/ProcessObserver.java:
```diff
@@ -137,7 +137,7 @@ public static void onEnterBackground() {
                 }
             } catch (Exception e) {
-                Logger.e(TAG, e.getMessage());
+                Logger.e(TAG, "Method onEnterBackground raised an exception: %s", e);
```
I'm just picking on this `Logger.e` as the message has changed in this PR but should we change all `Logger.e` to `Logger.track`?

AlexBenny replied:
As anticipated (sadly, weeks ago) it's not possible at the moment because the tracker doesn't allow service event and it can cause of infinite loops.
For now it's better limiting the error reporting for just few of them.

paulboocock commented on snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/tracker/Tracker.java:
```diff
@@ -568,6 +605,40 @@ public void run() {
 
     // --- Helpers
 
+    private void addServiceEventPayload(Payload payload, List<SelfDescribingJson> contexts, String eventId) {
```
Is a ServiceEvent an event from within the tracker that doesn't include any Subject information?
A method comment like addEventPayload would be useful to desribe how this differs from addEventPayload would be great.

AlexBenny replied:
Good idea!

AlexBenny commented:
@paulboocock  I've updated the code in order to track them internally rather than send them to the app.
It would be useful to send them to the app and to the collector, but for brevity (this PR really needs to be completed) I just report them to the collector and the user can see error in the logs.
I've also updated the documentation in the description of this PR. 

The associated schema is here: https://github.com/snowplow/iglu-central/pull/1023
